### PR TITLE
Boos 4345 gnn+llm graph reasoning ft

### PR DESCRIPTION
**Pull request by Johnno1011**
Just wanted to checkpoint and aggregate our work so here's a PR.

Added the following:
- Dataset class with collate function for batching.
- Batching GNN forward passes (thanks harvey <3)
- Batching LLM forward passes, allowing for EOS token or max new tokens to stop any sequence from completing early.
- Calculate loss based on cosine embeddings loss between openai and gllm response to a question (performs poorly)

### CODE REVIEW

File: src/primaite/agents/aegis/modules/llm.py [hunk 1]
```diff
@@ -80,100 +83,231 @@ def apply_chat_template(self, system: str, prompt: str, close_usr_msg: bool = Fa
...
+                embeds,
+            ],
+            dim=1,
+        )
+
+        # How many new tokens did we add to the beginning of the sequence?
+        n_tokens = embeds.shape[1] - original_len
+        attention_mask = torch.cat([torch.ones((batch_size, n_tokens)), attention_mask], dim=1)
+
+        # Expand all sequences to have minimum of close_msg_inputs tokens as slack
```
Comment (Johnno1011) [inline src/primaite/agents/aegis/modules/llm.py:117]: I did this before I knew that it doesn't matter about their being pad tokens in between bits of text - but no point undoing the work xD

File: src/primaite/agents/aegis/gllm.py [hunk 2]
```diff
@@ -39,105 +37,92 @@ def __init__(
...
-        return openai_tokens, gllm_tokens
+def train_loop(model: GLLM, dataloader: DataLoader, network_desc: str, n_epochs: int = 10):
+    model.train(mode=True)
 
+    for epoch in range(n_epochs):
+        gllm_responses = []
 
-def train_loop(model, x, edge_index, questions: List[str], network_desc: str):
+        for batch in dataloader:
+            model.optimizer.zero_grad()
```
Comment (HarveyAyling) [inline src/primaite/agents/aegis/gllm.py:100]: `zero_grad()` needs to be moved out of the batch loop or `model.optimizer.step()` moved in.

File: src/primaite/agents/aegis/gllm.py [hunk 3]
```diff
@@ -39,105 +37,92 @@ def __init__(
...
-        # Print gradients
-        for name, param in model.named_parameters():
-            print(f"Gradient of {name}: {param.grad}")
-        
         model.optimizer.step()
-        print(f"Episode {episode} Loss:", loss.item())
 
-    return gllm_responses, openai_responses
\ No newline at end of file
+        print(f"Epoch {epoch} Loss:", loss.item(), gllm_responses[0].replace("\n", "//n"))
```
Comment (HarveyAyling) [inline src/primaite/agents/aegis/gllm.py:116]: This only prints the loss of the last batch in the epoch - just to be aware.

Comment (Johnno1011) [reply src/primaite/agents/aegis/gllm.py:116]: Yes I am aware, I'm also printing only 1 response - it's just to give me a general idea of what it's saying :) Thanks

File: src/primaite/agents/aegis/gllm.py [hunk 4]
```diff
@@ -39,105 +37,92 @@ def __init__(
...
-            n_gat_layers=3,
+            n_tokens=100,
+            n_gat_layers=1,
             output_dim=self.llm.llm_embedding_size,
         ).to(self.ge_device)
-        self.openai = OpenAIClient(openai_api_key=self.openai_api_key)
 
-        #self.llm_optimizer = Adam(self.llm.parameters(), lr=self.llm_learning_rate)
+        # self.llm_optimizer = Adam(self.llm.parameters(), lr=self.llm_learning_rate)
         self.optimizer = Adam(self.parameters(), lr=self.ge_learning_rate)
```
Comment (HarveyAyling) [inline src/primaite/agents/aegis/gllm.py]: `llm_learning_rate` is not used anywhere :)

File: src/primaite/agents/aegis/gllm.py (hunk 2 above)
Comment (Johnno1011) [reply src/primaite/agents/aegis/gllm.py:100]: moved optimizer step in - good spot :D
